lib/services/access_request.go

```diff
@@ -245,53 +344,263 @@ func NewAccessRequest(user string, roles ...string) (AccessRequest, error) {
 	return &req, nil
 }
 
+func (c AccessRequestConditions) GetTraitMappings() TraitMappingSet {
+	tm := make([]TraitMapping, 0, len(c.ClaimsToRoles))
+	for _, mapping := range c.ClaimsToRoles {
+		tm = append(tm, TraitMapping{
+			Trait: mapping.Claim,
+			Value: mapping.Value,
+			Roles: mapping.Roles,
+		})
+	}
+	return TraitMappingSet(tm)
+}
+
 type UserAndRoleGetter interface {
 	UserGetter
 	RoleGetter
+	GetRoles() ([]Role, error)
 }
 
-func ValidateAccessRequest(getter UserAndRoleGetter, req AccessRequest) error {
-	user, err := getter.GetUser(req.GetUser(), false)
+// appendRoleMatchers constructs all role matchers for a given
+// AccessRequestConditions instance and appends them to the
+// supplied matcher slice.
+func appendRoleMatchers(matchers []parse.Matcher, conditions AccessRequestConditions, traits map[string][]string) ([]parse.Matcher, error) {
+	// build matchers for the role list
+	for _, r := range conditions.Roles {
+		m, err := parse.NewMatcher(r)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		matchers = append(matchers, m)
+	}
+
+	// build matchers for all role mappings
+	ms, err := conditions.GetTraitMappings().TraitsToRoleMatchers(traits)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return append(matchers, ms...), nil
+}
+
+// insertAnnotations constructs all annotations for a given
+// AccessRequestConditions instance and adds them to the
+// supplied annotations mapping.
+func insertAnnotations(annotations map[string][]string, conditions AccessRequestConditions, traits map[string][]string) {
+	for key, vals := range conditions.Annotations {
+		// get any previous values at key
+		allVals := annotations[key]
+
+		// iterate through all new values and expand any
+		// variable interpolation syntax they contain.
+	ApplyTraits:
```

Comment:
I don't think you need this label here, `continue` should continue to the next value in a inner loop.

Reply:
I try to always use loop labels if using `continue` within nested loops.  IMO its more readable and less prone subtle bugs due to later changes or bad merges.